### PR TITLE
[cronet_http] Enables CI for `cronet_http_embedded`

### PR DESCRIPTION
**Pull request by AlexV525**
Particially address #1047.

---

- [x] I’ve reviewed the contributor guide and applied the relevant portions to this PR.

<details>
  <summary>Contribution guidelines:</summary><br>

- See our [contributor guide](https://github.com/dart-lang/.github/blob/main/CONTRIBUTING.md) for general expectations for PRs.
- Larger or significant changes should be discussed in an issue before creating a PR.
- Contributions to our repos should follow the [Dart style guide](https://dart.dev/guides/language/effective-dart) and use `dart format`.
- Most changes should add an entry to the changelog and may need to [rev the pubspec package version](https://github.com/dart-lang/sdk/wiki/External-Package-Maintenance#making-a-change).
- Changes to packages require [corresponding tests](https://github.com/dart-lang/.github/blob/main/CONTRIBUTING.md#Testing).

Note that many Dart repos have a weekly cadence for reviewing PRs - please allow for some latency before initial review feedback.
</details>

### CODE REVIEW

File: .github/workflows/cronet.yml [hunk 1]
```diff
@@ -19,48 +21,45 @@ env:
...
-          java-version: '17'
-      - uses: subosito/flutter-action@v2
-        with:
-          channel: 'stable'
       - name: Run tests
         uses: reactivecircus/android-emulator-runner@v2
+        if: always() && steps.install.outcome == 'success'
         with:
           api-level: 28
           target: playstore
```
Comment (brianquinlan) [inline .github/workflows/cronet.yml:62]: Shouldn't this be false for cronet_embedded?

File: pkgs/cronet_http/android/build.gradle [hunk 2]
```diff
@@ -2,14 +2,14 @@ group 'io.flutter.plugins.cronet_http'
 version '1.0-SNAPSHOT'
 
 buildscript {
-    ext.kotlin_version = '1.6.10'
+    ext.kotlin_version = '1.7.21'
```
Comment (brianquinlan) [inline pkgs/cronet_http/android/build.gradle:5]: All of these types of upgrades should be done in a separate PR, no?

File: pkgs/cronet_http/example/android/app/build.gradle [hunk 3]
```diff
@@ -50,6 +41,7 @@ android {
         targetSdkVersion flutter.targetSdkVersion
         versionCode flutterVersionCode.toInteger()
         versionName flutterVersionName
+        multiDexEnabled true
```
Comment (brianquinlan) [inline pkgs/cronet_http/example/android/app/build.gradle:44]: This only needs to be set for cronet_http_embedded, right?

**Review by brianquinlan — COMMENTED** [review]
This is technically excellent work but I'm stuff not sure that `cronet_http_embedded` belongs in the same repo as `cronet_http`.

Let me talk to my team about it.

File: pkgs/cronet_http/android/build.gradle (hunk 2 above)
Comment (AlexV525) [reply pkgs/cronet_http/android/build.gradle:5]: I'm planning to reorg workflows and examples for the repo. These are necessary for the current setup. `1.6.10` is already outdated for the latest Flutter.

File: pkgs/cronet_http/example/android/app/build.gradle (hunk 3 above)
Comment (AlexV525) [reply pkgs/cronet_http/example/android/app/build.gradle:44]: Quite true, as it embeds more compared to using GMS's. But if we raise `minSdkVersion` to 21 we can also get rid of this.

File: .github/workflows/cronet.yml (hunk 1 above)
Comment (AlexV525) [reply .github/workflows/cronet.yml:62]: Not sure why we have `false` here. Can you explain more?

File: .github/workflows/cronet.yml [hunk 4]
```diff
@@ -19,48 +21,45 @@ env:
   PUB_ENVIRONMENT: bot.github
 
 jobs:
-  analyze:
-    name: Lint and static analysis
-    runs-on: ubuntu-latest
```
Comment (brianquinlan) [inline .github/workflows/cronet.yml:24]: Why did you change this to macos? I think that there is more linux machine availability.

File: .github/workflows/cronet.yml (hunk 1 above)
Comment (brianquinlan) [reply .github/workflows/cronet.yml:62]: Sorry, I meant that you probably do **not** want to use the playstore when testing cronet_http_embedded because it is designed to work without the playstore. I think that the value should be "default" for cronet_http_embedded. See https://github.com/ReactiveCircus/android-emulator-runner?tab=readme-ov-file#configurations

File: pkgs/cronet_http/android/build.gradle (hunk 2 above)
Comment (brianquinlan) [reply pkgs/cronet_http/android/build.gradle:5]: OK, feel free to set these to any reasonable values.

File: pkgs/cronet_http/example/android/app/build.gradle (hunk 3 above)
Comment (brianquinlan) [reply pkgs/cronet_http/example/android/app/build.gradle:44]: [Flutter supports all the way down to API 16](https://flutter-ko.dev/development/tools/sdk/release-notes/supported-platforms) but I some unofficial stats show that usage is pretty insignificant. So feel free to make the `minSdkVersion` 21 and get rid of this.

File: pkgs/cronet_http/example/android/app/src/main/java/io/flutter/app/FlutterMultiDexApplication.java [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+//
+// Modifications to this file should be done in a copy under a different name
+// as this file may be regenerated.
+
+package io.flutter.app;
+
+import android.app.Application;
+import android.content.Context;
+import androidx.annotation.CallSuper;
+import androidx.multidex.MultiDex;
```
Comment (brianquinlan) [inline pkgs/cronet_http/example/android/app/src/main/java/io/flutter/app/FlutterMultiDexApplication.java:13]: Remember that you can remove this too.

File: .github/workflows/cronet.yml (hunk 4 above)
Comment (AlexV525) [reply .github/workflows/cronet.yml:24]: There are unknown timeouts when running on the Ubuntu runner. https://github.com/dart-lang/http/actions/runs/7270023551/job/19808487300?pr=1070

File: pkgs/cronet_http/example/android/app/build.gradle (hunk 3 above)
Comment (AlexV525) [reply pkgs/cronet_http/example/android/app/build.gradle]: It's `19` now actually: https://github.com/flutter/flutter/blob/e86b8258196aeb7725b0dedc9f9b81f25547c47f/packages/flutter_tools/gradle/src/main/groovy/flutter.groovy#L46-L47

**Comment by brianquinlan** [comment]
If I add these transforms to `pkgs/cronet_http/tool/prepare_for_embedded.dart` then the tests pass. 

```dart

void updateImports() async {
  for (var file
      in _packageDirectory.listSync(recursive: true, followLinks: false)) {
    if (file.path.endsWith('.dart')) {
      final sourceFile = file as File;
      final updatedSource = sourceFile.readAsStringSync().replaceAll(
          'package:cronet_http/cronet_http.dart',
          'package:cronet_http_embedded/cronet_http_embedded.dart');
      sourceFile.writeAsStringSync(updatedSource);
    }
  }
}

void updateEntryPoint() async {
  File('${_packageDirectory.path}/lib/cronet_http.dart')
      .renameSync('${_packageDirectory.path}/lib/cronet_http_embedded.dart');
}
```

But the example application fails when making HTTP requests with:

```
E/flutter ( 7187): [ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: ClientException: Cronet exception: org.chromium.net.impl.NetworkExceptionImpl: Exception in CronetUrlRequest: net::ERR_CERT_DATE_INVALID, ErrorCode=11, InternalErrorCode=-201, Retryable=false, uri=https://www.googleapis.com/books/v1/volumes?q=tes&maxResults=40&printType=books
```

**Comment by AlexV525** [comment]
> But the example application fails when making HTTP requests with:
> 
> ```
> E/flutter ( 7187): [ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: ClientException: Cronet exception: org.chromium.net.impl.NetworkExceptionImpl: Exception in CronetUrlRequest: net::ERR_CERT_DATE_INVALID, ErrorCode=11, InternalErrorCode=-201, Retryable=false, uri=https://www.googleapis.com/books/v1/volumes?q=tes&maxResults=40&printType=books
> ```

That doesn't seem to be a library issue. I tried `flutter.cn` and it works fine. Sorry I cannot access `googleapis` from my location and the library doesn't seem to support a proxy.

**Comment by AlexV525** [comment]
> If I add these transforms to `pkgs/cronet_http/tool/prepare_for_embedded.dart` then the tests pass.

But the changes are treating. I'm going to add it to the current script.

**Comment by brianquinlan** [comment]
Hey @AlexV525,

Looks great! My intent is to merge this in the next couple of hours and then do a `package:cronet_http_embedded` release.

After that I'll increment the version number to 1.0 and do a `package:cronet_http` release but it will be functionally identical to the last `package:cronet_http_embedded` release.

We can work on switching publishers for `package:cronet_http_embedded` in the New Year. A lot of my colleagues are on vacation so I can't do it now.

Thanks so much for all your work on this, and on Dart HTTP in general!

**Review by brianquinlan — APPROVED** [review]

File: .github/workflows/cronet.yml [hunk 6]
```diff
@@ -19,48 +21,45 @@ env:
...
         uses: reactivecircus/android-emulator-runner@v2
+        if: always() && steps.install.outcome == 'success'
         with:
           api-level: 28
-          target: playstore
+          target: ${{ matrix.package == 'cronet_http_embedded' && 'default' || 'playstore' }}
           arch: x86_64
           profile: pixel
-          script: cd ./pkgs/cronet_http/example && flutter test --timeout=1200s integration_test/
+          script: cd 'pkgs/${{ matrix.package }}/example' && flutter test --timeout=1500s integration_test/
```
[the remark this answers is not in the document]
Comment (AlexV525) [reply .github/workflows/cronet.yml]: It doesn't change in the latest commit. I guess there are some other problems with the Ubuntu setup.

File: .github/workflows/cronet.yml [hunk 7]
```diff
@@ -6,10 +6,12 @@ on:
       - main
       - master
     paths:
+      - '.github/workflows/**'
```
[the remark this answers is not in the document]
Comment (AlexV525) [reply .github/workflows/cronet.yml:9]: Yeah sure. Will update these in another PR.

[the remark this answers is not in the document]
Comment (brianquinlan) [reply .github/workflows/cronet.yml:9]: I already updated this.
